Assist pilot to respect max speed and min height when relying on optical flow

priseborough opened the pull request:
Replaces https://github.com/PX4/Firmware/pull/8157.

This PR limits the max ground speed and minimum height above to respect limits published by the estimator during optical flow operation. The methods making these limits available were added to the ecl/EKF here: https://github.com/PX4/ecl/pull/339

These limits are only applied during manual controlled flight using the position controller.

The height control biases the vertical velocity demand to push the vehicle back up if below the minimum height. This bias is limited to be no more than the landing sink rate and can be easily over-powered by the pilot if required. It is intended to discourage sustained operation below a height where the flow sensor is in focus.

The position control limits the max speed to the value published by the estimator. This limit is released gradually if the estimator starts using GPS and is no longer reliant on the optical flow sensor to prevent sudden acceleration if GPS use commences while stick inputs are non-zero.

Flight test log here with GPS use commencing during flight: https://logs.px4.io/plot_app?log=2a18e88c-4068-4c76-967b-911f05993a94

SITL test log here with optical flow only operation : https://logs.px4.io/plot_app?log=fb4729b7-212e-4ef8-9144-d80e99c8ab00

priseborough commented:
The method of gradually releasing the speed limit when GPS i gained after taking off with just optical flow can still catch pilots by surprise out if they have been hovering for a while without control inputs. An alternative is once the speed  limiting has started, keep the limiting in place after GPS use has started, however allow it to increase with height gained as per normal, but not decrease when height is reduced again.

I will see how this method 'feels' from a user perspective.

priseborough commented:
Jenkins test this please

priseborough commented:
See https://github.com/PX4/ecl/pull/369 for description of testing at commit https://github.com/PX4/Firmware/pull/8223/commits/8b1dea134734af78d155639e62223caee5cc9a25

The flash limit for px4fmu-v2 is being exceeded.

priseborough commented:
Rebased

priseborough commented:
With the exception of the lack of flash space on the px4fmu-v2 build, the remaining CI failures have been fixed. I will now squash and clean up the commits.

priseborough commented:
Flight testing repeated at commit https://github.com/PX4/Firmware/pull/8223/commits/81f73271d0b482177d7542904f3059e1cf5fca3f

https://logs.px4.io/plot_app?log=866f61cd-965e-43ef-bc0e-034d43ce846e

@dagar px4fmu-v2 flash size limitation is failing jenkins CI

priseborough commented:
jenkins test this please

priseborough commented:
One of the ROS tests failed due to a preflight accel consistency check fail.

priseborough commented:
I've rebased and will re-test.

priseborough commented:
Logs uploaded here. Problems getting GPS to function with range finder and optical flow sensor. Further analysis to follow.

https://logs.px4.io/plot_app?log=70632044-ef2f-4cc8-ab6e-a2e7e7d21692
https://logs.px4.io/plot_app?log=186d9271-4779-4b36-ab11-7fe8cfa1dc73

priseborough commented:
GPS hardware was fixed and the combined use of optical flow and GPS tested:

https://logs.px4.io/plot_app?log=922ea875-5f4a-4d2a-9a26-8b4d006c28c1
https://logs.px4.io/plot_app?log=d1cca93e-ece6-442c-a6f2-79a5d833b689
https://logs.px4.io/plot_app?log=3e8f1eaf-e80b-4c39-bef3-28352602e255

GPS was unplugged for this test

https://logs.px4.io/plot_app?log=cafc8e65-6fc4-4a21-ad30-b9172374a882

Range finder performance was poor in these test which caused some unexpected height changes. Position hold was acceptable. Speed limiting and minimum height protection operated as expected when operating without GPS. Vehicle was able to arm and fly in position mode before GPS checks passed and gain GPS operation in flight.

priseborough commented:
Large number of range finder invalid readings and jump in height offset explain changes in height tracking during flight. The sensor used was an early model LL40LS silver label series :

![screen shot 2018-01-22 at 3 03 56 pm](https://user-images.githubusercontent.com/3596952/35204925-92a69126-ff85-11e7-8e33-855567c50641.png)

ChristophTobler commented:
@priseborough Thanks!
I quickly tried in simulation and works as it is implemented. But I don't see any reason for adding https://github.com/PX4/ecl/blob/master/EKF/ekf_helper.cpp#L1052-L1057. Also feels weird during flight. If I remove it, it behaves as expected.

priseborough commented:
Having the invalid data is helpful for review, however I would like to add min and max valid range parameters to the driver/sensor module. Will discuss this with you later.

priseborough commented:
@ChristophTobler https://github.com/PX4/ecl/blob/master/EKF/ekf_helper.cpp#L1052-L1057 was added in an attempt to handle the case where after flying around at low altitude without GPS and with speed limiting active, the vehicle then hovers for a few seconds and during that time GPs checks pass. The next time the pilot moves the stick by a large amount to reposition expecting the low sensitivity form stick->speed required for optical flow the copter reacts much more responsively causing flight into an adjacent object.

ChristophTobler commented:
@priseborough I see. How about we check if it changed from limiting to not limiting and gradually increase the limit?

priseborough commented:
If I understand you correctly that was what I had before, the problem being the limit was released gradually when the vehicle was in hover so the next stick movement gave a much larger response to the one before GPs was gained. What I will now do is make it so when limiting ceases, the limit is only raised at the specified rate when the input exceeds it.

ChristophTobler commented:
Yup, that's what I meant.:+1: Not after some fixed time but after stick input.

priseborough commented:
The speed limiter is not releasing after the last two patches. I will review the code and try to replicate on SITL.

priseborough commented:
Some jenkins builds are failing with the following error which is unrelated to the changes in this PR:

![screen shot 2018-01-24 at 4 51 04 pm](https://user-images.githubusercontent.com/3596952/35316638-e3708af6-0126-11e8-9e99-2e5a5e41a231.png)

dagar commented:
It's brought in by the ecl update, and for some reason only certain compiler versions seem to mind. https://github.com/PX4/ecl/compare/cdbca91e79fad33a8908c009d66f14a182d55f1e...3fc60290b702f574b79b5ef61b7e8b32b1cf2829#diff-2102c0e5cfe8ea379e7f50690ee77b52R73

Cody-Chen commented:
@priseborough Is possible to disable flow fuse when the drone has gps lock?

priseborough commented:
@cody, yes it is, however I will make it a parameter selectable option because the position hold performance with poor GPS is superior when also using optical flow. The EKF will not fuse flow data if quality metric is poor or flow level is out of range.

Cody-Chen commented:
Thansk. If have a parameter to select use flow fusion or not.  I will prefer to use EKF2_AID_MASK to enable flow fusion as default. Once the flow module publish any notice (like in dark environment). We just  disable flow fusion.

priseborough commented:
Reverting recent data validator changes has fixed the build error, but we are getting a failure on the ROS VTOL mission.

dagar commented:
The VTOL mission tests are still failing intermittently (https://github.com/PX4/Firmware/issues/8556). Restarting.

ChristophTobler commented on src/modules/mc_pos_control/mc_pos_control_main.cpp:
```diff
@@ -3048,8 +3089,15 @@ MulticopterPositionControl::task_main()
+		}
+
+		/* never exceed the parameter specified maximum */
+		_vel_max_xy = fminf(_vel_max_xy, _params.vel_max_xy);
```
@priseborough That part doesn't work. `_vel_max_xy` is initialized to `0.0f` and will stay very small.
Try to run `make posix_sitl_default gazebo` and fly in position control. The quad will barely move.

ChristophTobler replied:
because `_vel_sp`/`vel_norm_xy` is bound to `_vel_max_xy` https://github.com/PX4/Firmware/blob/pr-ekfPosCtrlLimits/src/modules/mc_pos_control/mc_pos_control_main.cpp#L1393

priseborough replied:
That's strange i didn't get that result when I flew outdoors earlier, but it is definitely doing it on SITL. Will fix.

priseborough commented:
@ChristophTobler I've fixed the bug and cleaned up the commit history.  It has been tested with 'make posix_sitl_default gazebo' and in position mode and 'make posix gazebo_iris_opt_flow' in position mode. With the optical flow simulation I enabled GPS use after takeoff by changing EKF2_AID_MASK from 2 to 3 and verified the speed limit was released as expected.

I am working on another project tomorrow, but will be able to get some further testing on Friday.

dagar commented:
Good to merge?

ChristophTobler commented:
@dagar I think so. @priseborough?

priseborough commented:
Yes, good to go.
